internal/template-bundle/bundle.go

```diff
@@ -103,18 +61,32 @@ func extractDataSources(templates []templatev1.Template) ([]cdiv1beta1.DataSourc
 		}
 
 		namespace, exists := findDataSourceNamespace(template)
-		if !exists {
-			continue
+		// This check is needed, so later code can assume that all DataSources
+		// should be created in the internal.GoldenImagesNamespace
+		if exists && namespace != internal.GoldenImagesNamespace {
+			// If this happens, it is a programmer's error.
+			return nil, fmt.Errorf(
+				"common template %s has invalid default DATA_SOURCE_NAMESPACE value: %s, expected: %s",
+				template.Name, namespace, internal.GoldenImagesNamespace)
 		}
 
-		namespacedName := namespace + "/" + name
-		if _, duplicateName := uniqueNames[namespacedName]; !duplicateName {
-			dataSources = append(dataSources, createDataSource(name, namespace))
-			uniqueNames[namespacedName] = struct{}{}
+		if dataSourceArchs[name] == nil {
+			dataSourceArchs[name] = map[string]struct{}{}
+		}
+
+		templateArch := common_templates.GetTemplateArch(template)
+		dataSourceArchs[name][templateArch] = struct{}{}
+	}
+
+	result := make(map[string][]string, len(dataSourceArchs))
+	for name, archs := range dataSourceArchs {
+		result[name] = make([]string, 0, len(archs))
+		for arch := range archs {
+			result[name] = append(result[name], arch)
```

Comment:
Why go the extra route by first collecting arches into a map and then creating a slice from it? You could collect them into a slice and sort/compact it after. This is very hard to understand IMO.

Reply:
done